common/test/TestLogging.cs

```diff
@@ -2,39 +2,116 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
-using System.Diagnostics;
-using System.Globalization;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Versioning;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
 
 namespace Microsoft.Azure.Devices.E2ETests
 {
+    /// <summary>
+    /// This class logs to multiple providers - Event source, Application insights
+    /// </summary>
     public class TestLogging
     {
-        private readonly static TestLogging s_instance = new TestLogging();
+        public const string Language = "C#";
+        public const string Service = "IotHub";
+        private readonly object initLock = new object();
+
+        // Client to log to application insights
+        private static TelemetryClient _telemetryClient;
+
+        // Unique ID for all tests of a run.
+        private static string _testRunId;
+
+        public TestContext Context { get; set; }
+        public string TargetFramework { get; set; }
 
         private TestLogging()
         {
+            if (_telemetryClient != null)
+            {
+                return;
+            }
+            else
+            {
+                // Thread-safe initialization of the telemetry client and run Id.
+                lock (initLock)
```

Comment:
there needs to be another check for `_telemetryClient != null` inside this lock.

If we have 2 tests accessing this at the same time, both will evaluate `_telemetryClient != null` as false, and try to grab the lock.
suppose test1 succeeds, it will grab the lock, initialize `_telemetryClient`. Once the lock is released, test2 will grab it again, still think `_telemetryClient` is null, and try to reinitialize.

Reply:
yes you are right. I had it that way and wanted to optimize but this does not work. I guess we have no other way than locking always.